### PR TITLE
feat: refactor s3 configurations

### PR DESCRIPTION
**Pull request by frankli0324**
breaking quite a lot of stuff here

* dev compose file needs a rewrite
* previous deployments would have to be reconfigured

however it (aims to) bring:

* compatibility with more S3 providers
* sanity

### CODE REVIEW

**Comment by phongkt-dev** [comment]
hmm, I'm using both Minio and AWS S3, and have no problem with current version of Outline, why we need to change env variable name?

**Comment by frankli0324** [comment]
I find the current implementation quite confusing and seems to be suffering from burdens of the past.
look at this, just... how?

https://github.com/outline/outline/blob/7f15eb287debce76529a50625901ee1933498968/server/utils/s3.ts#L132-L145

also, I'm not expecting this PR to be merged soon. I changed the variable names because I expect to find a way to add backward compatibility, by "inferring" the added variables from the old variable values.

https://github.com/outline/outline/discussions/3159

**Comment by tommoor** [comment]
This is a great initiative

File: server/presenters/document.ts [hunk 1]
```diff
@@ -20,7 +20,10 @@ async function replaceImageAttachments(text: string) {
         const signedUrl = await getSignedUrl(attachment.key, 3600);
         text = text.replace(
           new RegExp(escapeRegExp(attachment.redirectUrl), "g"),
-          signedUrl
+          // keep "attachments.redirect" string for files in shared urls to be recognized correctly
+          // this behavior should be changed upon switching to `json` based document transfer
+          //     in `documents.info` endpoint
+          signedUrl + "# attachments.redirect"
```
Comment (tommoor) [inline server/presenters/document.ts:26]:
```suggestion
          signedUrl + "#attachments.redirect"
```

Is there a reason for the space here?

**Comment by tommoor** [comment]
> is there any other changes that must be applied?

It's honestly difficult to tell as there are so many different deployments out there for this particular part of the config. It might be best to put it out as a patch release or something before the next minor to try and gather any bugs

File: server/utils/s3.ts [hunk 2]
```diff
@@ -1,110 +1,66 @@
...
-import { addHours, format } from "date-fns";
 import fetch from "fetch-with-proxy";
 import { v4 as uuidv4 } from "uuid";
 import Logger from "@server/logging/logger";
 
+// backward compatibility
 const AWS_S3_ACCELERATE_URL = process.env.AWS_S3_ACCELERATE_URL;
+const AWS_S3_UPLOAD_BUCKET_URL = process.env.AWS_S3_UPLOAD_BUCKET_URL;
+const AWS_S3_UPLOAD_BUCKET_NAME = process.env.AWS_S3_UPLOAD_BUCKET_NAME;
+const AWS_S3_FORCE_PATH_STYLE = process.env.AWS_S3_FORCE_PATH_STYLE;
```
Comment (tommoor) [inline server/utils/s3.ts:11]: Thoughts on logging out a warning of deprecation and future removal if these are used?

**Comment by tommoor** [comment]
Thanks for the ongoing work, I just tested locally with my existing config – which looks like:

```
AWS_REGION=us-east-2
AWS_S3_ACCELERATE_URL=https://outline-staging-attachments.s3-accelerate.amazonaws.com
AWS_S3_FORCE_PATH_STYLE=false
AWS_S3_UPLOAD_BUCKET_NAME=outline-staging-attachments
AWS_S3_UPLOAD_BUCKET_URL=https://outline-staging-attachments.s3.us-east-2.amazonaws.com
AWS_S3_UPLOAD_MAX_SIZE=26214400
```

This resulted in an upload failure, it seems like the bucket name ends up getting doubled on the request:

image

**Comment by frankli0324** [comment]
hmmmm... may I just ...`.replace(`${bucket_name}.`, '');`?

**Comment by tommoor** [comment]
Maybe it's the only way 😢

File: server/utils/s3.ts [hunk 3]
```diff
@@ -1,110 +1,86 @@
...
-import { addHours, format } from "date-fns";
 import fetch from "fetch-with-proxy";
 import { v4 as uuidv4 } from "uuid";
 import Logger from "@server/logging/logger";
 
-const AWS_S3_ACCELERATE_URL = process.env.AWS_S3_ACCELERATE_URL;
+// backward compatibility
+function removeBucketName(url: string | undefined) {
+  if (
+    !process.env.AWS_S3_FORCE_PATH_STYLE &&
```
Comment (tommoor) [inline server/utils/s3.ts:10]:
```suggestion
    process.env.AWS_S3_FORCE_PATH_STYLE !== "true" &&
```

**Comment by tommoor** [comment]
Once this merges I'm going to do a big centralization and refactor of env handling, that will help a lot.

**Comment by redraskal** [comment]
> is there any other changes that must be applied?

The ACL option in presigned posts should be optional.

1) Amazon says it is https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html
2) Not every s3 alternative supports it, mine does not support ACL and rejects attachment uploads

This ties in with https://github.com/outline/outline/issues/2769

**Comment by NK74** [comment]
Any updates ?

**Comment by almereyda** [comment]
As far as I remember, @tommoor still considered this a useful improvement over the current way of configuring S3. Is there a chance this can be reopened, and have a timely (in the sense of the bot) review?

**Comment by frankli0324** [comment]
I'm very sorry but I simply don't have the effort and motivation to work on this after I've graduated and start working...
I'll still try to finish this up if no one else do

**Comment by Hoshino-Yumetsuki** [comment]
It's 2025 and Outline still doesn't seem to have made any improvements to S3 API compatibility, and hopefully this pull request will be merged

**Pull request closed without merging**